src/main/scala/diplomacy/Nodes.scala

```diff
@@ -5,79 +5,238 @@ package freechips.rocketchip.diplomacy
+
+  /** Create an inward Bundle parameterized by the inward edge
+    *
+    * @param ei inward Edge of this node.
+    * @return inward Bundle of this node
+    */
   def bundleI(ei: EI): BI
 
-  // Edge functions
+  /** function to generate and attach a monitor for this node input.
+    *
+    * @param bundle input bundle of this node to attach the monitor to
+    * @param edge edge of this node used to parameterize the bundle
+    * */
   def monitor(bundle: BI, edge: EI) {}
+
+  /** Define how the edge should be rendered (e.g. in GraphML)
+    *
+    * @param e edge to render
+    * @return [[RenderedEdge]] description of how the edge should be generated
+    */
   def render(e: EI): RenderedEdge
 
-  // optional methods to track node graph
-  def mixI(pu: UI, node: InwardNode[DI, UI, BI]): UI = pu // insert node into parameters
+  /** Defines how input parameters can be "mixed" or negotiated together.
+    *
+    * The default behavior is to just return `pu`.
+    *
+    * @param pu   upward flowing edge parameters
+    * @param node an inward node to "mix" the upward flowing parameters into
+    * @return altered version of the upward flowing inward parameters
+    */
+  def mixI(pu: UI, node: InwardNode[DI, UI, BI]): UI = pu
 }
 
-// DO = Downwards flowing Parameters generated by the outer side of the node
-// UO = Upwards   flowing Parameters received on the outer side of the node
-// EO = Edge Parameters describing a connection on the outer side of the node
-// BO = Bundle type used when connecting to the outer side of the node
+/** [[OutwardNodeImp]] defines the types that describe the outwards side of the [[BaseNode]].
+  *
+  * @tparam DO The type of the downwards flowing parameters generated by the outer side of the node
+  * @tparam UO Tye type of the upwards flowing parameters received by the outer side of the node
+  * @tparam EO The type of the edge parameters (after diplomatic resolution) describing a connection on the outer side of the node.
+  * @tparam BO The type of the Bundle used when connecting to the outer side of the node, corresponding to the real hardware interface that is emitted along the graph edge, generally paramterized by the EO type.
+  */
 trait OutwardNodeImp[DO, UO, EO, BO <: Data]
 {
+  /** Create an outward-facing edge on this [[BaseNode]]
+    *
+    * @param pd         the downward flowing parameters out of the node along the edge
+    * @param pu         the upward-flowing parameters going into the node along the edge
+    * @param p          [[Parameters]] accessible by this node for whatever purpose
+    * @param sourceInfo [[SourceInfo]] of this edge
+    * @return an outward edge of this node
+    */
   def edgeO(pd: DO, pu: UO, p: Parameters, sourceInfo: SourceInfo): EO
+
+  /** Create an outward Bundle parameterized by the inward edge
+    *
+    * @param eo the outward Edge of this node
+    * @return an outward Bundle of this node
```

Comment:
`an outward Bundle of this node parameterized by the negotiated Edge parameters`